docs/tutorials/forum-application/3.send-message.md

```diff
@@ -0,0 +1,381 @@
+or the new hash of the application state.
+
+**Following is the code for the `FinalizeBlock` function:**
+
+```go
+// Deliver the decided block to the Application
+func (app *ForumApp) FinalizeBlock(_ context.Context, req *abci.FinalizeBlockRequest) (*abci.FinalizeBlockResponse, error) {
+	fmt.Println("Executing Application FinalizeBlock")
+
+	// Iterate over Tx in current block
+	app.onGoingBlock = app.state.DB.GetDB().NewTransaction(true)
+	respTxs := make([]*abci.ExecTxResult, len(req.Txs))
+	finishedBanTxIdx := len(req.Txs)
+	for i, tx := range req.Txs {
+		var err error
+
+		if isBanTx(tx) {
+			banTx := new(model.BanTx)
+			err = json.Unmarshal(tx, &banTx)
+			if err != nil {
+				respTxs[i] = &abci.ExecTxResult{Code: CodeTypeEncodingError}
+			} else {
+				err := UpdateOrSetUser(app.state.DB, banTx.UserName, true, app.onGoingBlock)
+				if err != nil {
+					panic(err)
+				}
+				respTxs[i] = &abci.ExecTxResult{Code: CodeTypeOK}
+			}
+		} else {
+			finishedBanTxIdx = i
+			break
+		}
+	}
+
+	for idx, tx := range req.Txs[finishedBanTxIdx:] {
+		// From this point on, there should be no BanTxs anymore
+		// If there is one, ParseMessage will return an error as the
+		// format of the two transactions is different.
+		msg, err := model.ParseMessage(tx)
+		i := idx + finishedBanTxIdx
+		if err != nil {
+			respTxs[i] = &abci.ExecTxResult{Code: CodeTypeEncodingError}
+		} else {
+			// Check if this sender already existed; if not, add the user too
+			err := UpdateOrSetUser(app.state.DB, msg.Sender, false, app.onGoingBlock)
+			if err != nil {
+				panic(err)
+			}
+			// Add the message for this sender
+			message, err := model.AppendToExistingMsgs(app.state.DB, *msg)
+			if err != nil {
+				panic(err)
+			}
+			app.onGoingBlock.Set([]byte(msg.Sender+"msg"), []byte(message))
+			chatHistory, err := model.AppendToChat(app.state.DB, *msg)
+			if err != nil {
+				panic(err)
+			}
+			// Append messages to chat history
+			app.onGoingBlock.Set([]byte("history"), []byte(chatHistory))
+			// This adds the user to the DB, but the data is not committed nor persisted until Commit is called
+			respTxs[i] = &abci.ExecTxResult{Code: abci.CodeTypeOK}
+			app.state.Size++
+		}
+	}
+	app.state.Height = req.Height
+
+	response := &abci.FinalizeBlockResponse{TxResults: respTxs, AppHash: app.state.Hash()}
+	return response, nil
+}
+```
+
+**Explanation of code:**
+
+`FinalizeBlock` function iterates over the transactions (`req.Txs`) in the block. If a transaction is a ban transaction (`isBanTx(tx)`),
+it updates or sets a user in the application's database and assigns a response code. If it is not a ban transaction, the loop breaks.
+
+After the ban transactions, the method parses the remaining transactions using `model.ParseMessage(tx)`. It updates or sets
+the sender as a user in the database, appends the message to the existing messages, appends the message to the chat history,
```

Comment:
> appends the message to the existing messages

Why do we do that (trying to understand)?

Reply:
I think the way the logic to have Ban Tx and "regular" Tx could be improved in the future imo. I'd leave like that for now, might not be optimal but I'd consider this as an future enhancement.